### PR TITLE
feat!: support multiple parallel `child_process`

### PR DESCRIPTION
**Pull request by AriPerkkio**
### Description

- Many projects are unable to use `node:worker_threads` and are forced to use a single `node:child_process` via `--no-threads`. That's stable, but slow. This PR allows using multiple `node:child_process` for running tests similarly as `--threads` does.
  - https://github.com/vitest-dev/vitest/issues/3143
  - https://github.com/vitest-dev/vitest/issues/3106
  - https://github.com/vitest-dev/vitest/issues/3816
  - https://github.com/tinylibs/tinypool/issues/48
- Ref. https://github.com/tinylibs/tinypool/issues/64

#### BREAKING CHANGES

`useFakeTimers()` no longer automatically mocks `process.nextTick`. Users can still mock it by explicitly specifying `useFakeTimers({ toFake: ['nextTick'] })`.

Mocking `process.nextTick` by using `useFakeTimers({ toFake: ['nextTick'] })` is no longer supported when running in `--no-threads`. Node's `child_process` uses `process.nextTick` internally and it is really difficult to support this use case. Users can still mock `process.nextTick` when running with `--threads`.

`--no-threads` no longer runs tests in the same test context. Tests are now isolated in separate processes similar to how `--threads` does with `worker_threads`.

#### Testing

 - [x] `pnpm test:ci:no-threads`
 - Manual testing with third party projects:
   - [x] [`zustand`](https://github.com/pmndrs/zustand): Tests require isolation so comparing speed change doesn't make sense here. All tests pass with new isolated `--no-threads`.
   - [x] [`vuejs/core`](https://github.com/vuejs/core). Uses `--no-threads` on CI. Speed improvement 48.57s -> 26.96s.
   - [x] [`netlify/cli`](https://github.com/netlify/cli). Tests pass with `--no-threads`.
   - [x] [`lerna-lite`](https://github.com/lerna-lite/lerna-lite). Tests require `child_process` and have been using `--no-threads`. Speed improvement 67s -> 24s.

___

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by AriPerkkio** [comment]
Seeing some process hangs when running Vitest's test cases with `--no-threads`. Debugging these is really time consuming and I've spent quite some time looking into this. I'm starting to think we should implement graceful `worker_thread` and `child_process` termination in Tinypool. That would allow us to implement Jest's `--detect-open-handles` feature too.

These errors come up locally and on CI:

```
⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯
Error: [birpc] timeout on calling "onTaskUpdate"
 ❯ Timeout._onTimeout ../../packages/vitest/dist/vendor-index.9acf798a.js:38:22
 ❯ listOnTimeout node:internal/timers:569:17
 ❯ processTimers node:internal/timers:512:7
```

**Comment by sheremet-va** [comment]
> Debugging these is really time consuming and I've spent quite some time looking into this. 

Welcome to my world! For me it's almost always something related to timer mocks (nextTick included).

> I'm starting to think we should implement graceful worker_thread and child_process termination in Tinypool.

Will it help? Or just to debug?

**Comment by AriPerkkio** [comment]
> Will it help? Or just to debug?

Mostly for debugging. But still, I'm not sure if it's good practice to terminate workers and processess with `.kill()` and `.terminate()` when there can still be all kind of stuff running, e.g. chokidar with fsevents. I'm starting to think that it's causing some of these process hangs but don't really have proof for it.

The way how `jest-worker` does this is that main thread instructs thread/process (worker) to [remove all its message handlers](https://github.com/jestjs/jest/blob/25a8785584c9d54a05887001ee7f498d489a5441/packages/jest-worker/src/workers/processChild.ts#L155C52-L156) so that there is nothing keeping the worker running, and then it waits for the worker to gracefully exit. If it doesn't exit in specific timeout, main threads sends it SIGTERM. If it still doesn't exit, [it sends SIGKILL](https://github.com/jestjs/jest/blob/25a8785584c9d54a05887001ee7f498d489a5441/packages/jest-worker/src/workers/ChildProcessWorker.ts#L463C8-L477). What Tinypool currently does is that it immediately sends SIGTERM and then hangs if worker doesn't exit - or throws [timeout error](https://github.com/tinylibs/tinypool/pull/50).

But if Vitest used graceful termination for workers, it could impact in performance. We would not be killing workers as fast as possible - instead we would be waiting for them to shut down themselves. I'm not sure if this performance issue is real or just speculation. Jest or Vitest's `vm-threads` don't really suffer from this as isolation doesn't require recycling workers.

Maybe I'll implement `--detect-open-handles` into Tinypool so that it's optional flag that can be turned on when needed. It would definitely help debugging this PR's issues.

File: test/core/test/fixtures/timers.suite.ts [hunk 1]
```diff
@@ -338,7 +338,7 @@ describe('FakeTimers', () => {
     })
   })
 
-  describe('runAllTimersAsync', () => {
+  describe.skipIf('TODO Enable this')('runAllTimersAsync', () => {
```
Comment (AriPerkkio) [inline test/core/test/fixtures/timers.suite.ts:341]: > For me it's almost always something related to timer mocks

To talk about timers... 😅 
Every other test is now passing (even on CI) but this one. This is throwing the `Error: [birpc] timeout on calling "onTaskUpdate"` error. I think that error used to come up on `main` branch's CI maybe a month ago but now it's gone from there.

I can reproduce this error locally on MacOS when using [`stress(1)` CLI tool](https://github.com/resurrecting-open-source-projects/stress) to simulate CPU load. That tool is available on `brew` for easy installation.

```
# In one terminal
$ stress --cpu 8 --io 4 --vm 2 --vm-bytes 128M --timeout 200s

# In second terminal. Run couple of times if it doesn't come up immediately.
$ cd vitest/test/core 
$ pnpm run test --run --no-threads --allow-only
```

I've been looking at Vitest's `onTaskUpdate` usage for long now and cannot understand why that error is coming up. Runner should be `await`'ing that call to finish. But I guess when timers are mocked, all kinds of unexpected behavior can come up. The test case is even passing `global.Promise` to `fake-timers`. 🤔 

I think I'll take a look into `@sinonjs/fake-timers` sources next.

Comment (AriPerkkio) [reply test/core/test/fixtures/timers.suite.ts]: Looks like this is actually a bug on `main` and doesn't exactly relate to this PR's changes.

When a test case mocks `process.nextTick` during test case, any execution that relies on `nextTick` will not be run. When `useRealTimers()` is called, those pending tasks are still not executed. In `birpc` we are storing the original `setTimeout`, `clearTimeout` and `Math.random` to prevent similar cases. I'm not sure if we even can keep internal dependencies running properly when test execution shares the global context with runner. VM isolation might help here?

Here is test case that will fail on latest Vitest release or `main` branch when run as `vitest --no-threads`. It might need to be ran multiple times, or slightly increase the timeout value. 

```ts
import { test, vi } from "vitest";

test("leaky process mocking", async () => {
  const ogSetTimeout = setTimeout;

  // Fake `process.nextTick` and execute longer than 10ms. Use heavy looping or simple `setTimeout`
  vi.useFakeTimers({ toFake: ["nextTick"] });

  process.stdout.write("\nThis is executed\n\n");
  await new Promise((resolve) => ogSetTimeout(resolve, 100)); // onTaskUpdate uses 10ms
  process.stdout.write("\nThis is also executed\n\n");

  vi.useRealTimers();

  process.stdout.write("\nTest finished\n\n");
});
```

The test case runs for 60s and `birpc` timeouts. Also it fails to report failed test case from child process now. 

image



Here is minimal-ish reproduction using just `@sinonjs/fake-timers` and `node:child_process`. When main process uses `process.send`, and subprocess has `process.nextTick` mocked, the `process.on("message", callback)` is not executed. And after `useRealTimers()` call those pending tasks are not executed at all.

<details>

```js
/*
 * Sends multiple messages from main process to subprocess.
 * Subprocess will fail if it does not receive all the messages.
 * Between the message sending, subprocess will install fake-timers.
 * After a while, it will uninstall fake-timers.
 *
 * Requires:
 * pnpm i @sinonjs/fake-timers@11.0.0
 */

import { fork } from "node:child_process";
import { fileURLToPath } from "node:url";

const filename = fileURLToPath(import.meta.url);
const isMainThread = !Boolean(process.env.CHILD_PROCESS);

const INTERVAL = 1000; // Time in ms between messages
const LOOPS = Array(10).fill(0).map((_, i) => i); // Amount of messages to be sent

if (isMainThread) {
  console.log("[main] running");
  const subprocess = fork(filename, { env: { CHILD_PROCESS: "true" } });
  const onExit = new Promise((r) => subprocess.on("exit", r));
  const onReady = new Promise((r) => subprocess.on("message", (m) => m === "ready" && r()));

  await onReady;
  console.log("[main] subprocess ready\n");

  // Send amount of LOOPS to subprocess with INTERVAL delay in between
  for (const index of LOOPS) {
    await new Promise((r) => setTimeout(r, INTERVAL));
    console.log("[main] sending message", index);
    subprocess.send({ loop: index });
  }

  // Let process exit when it's done
  subprocess.send({ exit: "graceful" });

  await onExit;
  console.log("[main] subprocess exited");
} else {
  console.log("[subprocess] running");

  const ogSetTimeout = setTimeout;
  const ogClearTimeout = clearTimeout;

  // These are used to indicate whether all messages were not received
  const timeouts = LOOPS.map((i) =>
    ogSetTimeout(() => { throw new Error(`Timeout ${i}`) }, INTERVAL * LOOPS.length * 1.5)
  );

  async function onMessage(message) {
    // Remove timeout from given LOOP message. If any messages are missed, a timeout error is expected
    if (message?.loop !== undefined) {
      console.log("[subprocess] received loop", message.loop);
      ogClearTimeout(timeouts[message.loop]);
      return;
    }

    if (message?.exit === "graceful") {
      process.off("message", onMessage);
      return console.log("[subprocess] exitting gracefully");
    }

    console.log("[subprocess] received unknown message", message);
  }

  process.on("message", onMessage);

  // Setup fake-timers
  const { withGlobal } = await import("@sinonjs/fake-timers");

  class FakeDate extends Date {}
  const global = { Date: FakeDate, clearTimeout, process, setTimeout, Promise };
  const fakeTimers = withGlobal(global);

  // Install fake-timers after 3s. After another 3s uninstall them.
  ogSetTimeout(() => {
    console.log("[subprocess] installing fake-timers");
    const clock = fakeTimers.install({
      now: Date.now(),
      toFake: Object.keys(fakeTimers.timers),
    });

    setTimeout(() => {
      console.log("[subprocess] uninstalling fake-timers");
      clock.uninstall();
    }, 3000);
  }, 3000);

  process.send("ready");
}
```

</details>

Does this analysis sound correct? The amount of debugging across `vitest`, `birpc`, `tinypool` and `@sinonjs/fake-timers` was a lot. 😫

Comment (sheremet-va) [reply test/core/test/fixtures/timers.suite.ts]: Sounds correct, I also had a lot of problems with `nextTick` the first time child process wad implemented. I fixed it by restoring original `nextTick` before calling rpc method, does it not work anymore?

Comment (AriPerkkio) [reply test/core/test/fixtures/timers.suite.ts]: > fixed it by restoring original `nextTick` before calling rpc method

Good point, I was unaware of this. This is done by `withSafeTimers`, right?

That one is not `await`'ing the given callback. Calls of `birpc` are Promises when they are not specified in its `eventNames` config. By adding `await` here the failing minimal test case above is constantly passing:

https://github.com/vitest-dev/vitest/blob/91853bb204ed8dfc3750d54b1b79409168ea440b/packages/vitest/src/runtime/rpc.ts#L30-L31

However that one still doesn't fix all issues, as now `rpc().snapshotSaved` seems to timeout on Windows. I'll continue debugging. 🙃

Comment (sheremet-va) [reply test/core/test/fixtures/timers.suite.ts]: The main reason why it wasn’t awaiting is because the function itself was running synchronously and returned a promise, so all globals should’ve been called by that time. Maybe implementation changed in the meantime? Or maybe it was never working as intended 👀

Comment (AriPerkkio) [reply test/core/test/fixtures/timers.suite.ts]: > function itself was running synchronously and returned a promise, so all globals should’ve been called by that time.

When the rpc method is called synchronously, it means that only the subprocess's `process.send` is ensured to be run using original `nextTick`. The response coming from main process is not received in `process.on('message', callback)` if `nextTick` is mocked. We'll need to `await` the whole rpc call to make sure response is received from main process. 

But this also means that any code in user's test case that runs between this cycle is running without mocked `nextTick`.

Comment (sheremet-va) [reply test/core/test/fixtures/timers.suite.ts]: I would just disallow mocking `nextTick` to be honest

Comment (AriPerkkio) [reply test/core/test/fixtures/timers.suite.ts]: I'm starting to think that would be the best option too. This is driving me crazy. Debugging NodeJS when such low level internals are mocked is not fun. 
We could raise an error if `useFakeTimers({toFake: ['nextTick']})` is called when `node:child_process` is used. I don't think that many users are even interested in `nextTick` mocking.

**Comment by ghiscoding** [comment]
> * [`lerna-lite`](https://github.com/lerna-lite/lerna-lite). Tests require `child_process` and have been using `--no-threads`. Speed improvement 67s -> 24s. Some snapshots are wrapped in double quotes now. 🤔

@AriPerkkio I'm really excited about this PR, especially that I'm on Windows so this will be a big boost for me... and I just saw that you started testing Lerna-Lite, awesome, so about the comment you wrote, I might be able to shed some light, I had a few issues when I first moved to Vitest (like this issue #3129 that I logged in Vitest). Basically Lerna (Lerna-Lite is a fork of Lerna) had many custom serializers and running them with `--no-threads` had the consequence of being run in the same global scope causing some problems as described by @sheremet-va in the referenced issue 

> All tests with the same environment in `--no-threads` runs in a single child_process and they all share the same global scope. Is it possible that one test just affects the other?

I fixed this problem in this Lerna-Lite [PR](https://github.com/lerna-lite/lerna-lite/pull/650) by duplicating some of these serializer in certain spec files that had quote escaping issues

So in short, it might be related to the double quotes problem you are encountering when testing Lerna-Lite. I don't mind reverting this Lerna-Lite [PR](https://github.com/lerna-lite/lerna-lite/pull/650) or changing any code that could help with the extra quotes you're seeing, whatever makes it works, I'm all good with it 😉

**Comment by AriPerkkio** [comment]
Thanks for the insight @ghiscoding - you just save me an hour or two of debugging. In the current testing setup of `lerna-lite` the snapshot serializers are indeed leaking from one test file to another. I fixed the snapshot issues by adding the serializer to following files: `version-build-metadata.spec.ts`, `version-conventional-commits.spec.ts`, `version-ignore-changes.spec.ts`. Now all tests are passing and running quickly.

**Comment by AriPerkkio** [comment]
What kind of API should `child_process` pool expose to users, @sheremet-va?

- `--single-process`, `minProcess`, `maxProcess` (equivalents of `--single-thread`, `minThreads`, `maxThreads`
- Or simply call threads and processes as "workers" as Jest does, and have `--single-worker`, `minWorkers`, `maxWorkers`? Pool selection would be done using `--threads`, `--no-threads`? 

Currenly this PR does not implement a `child_process` equivalent of `--single-threads` - do we need one?

**Comment by sheremet-va** [comment]
The problem here is naming 😄

Maybe implement current names for bow, but we already discussed the new naming convention for pools - maybe we can implement it in the same breaking change release, but in a separate PR.

**Comment by ghiscoding** [comment]
> Thanks for the insight @ghiscoding - you just saved me an hour or two of debugging. In the current testing setup of `lerna-lite` the snapshot serializers are indeed leaking from one test file to another. I fixed the snapshot issues by adding the serializer to following files: `version-build-metadata.spec.ts`, `version-conventional-commits.spec.ts`, `version-ignore-changes.spec.ts`. Now all tests are passing and running quickly.

I would be glad to receive a PR with all your changes for this upcoming feature, even if it's in draft until this is released, that would save me some time when it becomes available. Thanks a ton for all your hard work :)

**Comment by AriPerkkio** [comment]
> Maybe implement current names for now, but we already discussed the new naming convention for pools - maybe we can implement it in the same breaking change release, but in a separate PR.

Yes, now I remember it - again. 😄 Configuring via pools make this much easier. 

I'll leave this PR to use `--no-threads` and implement `--single-thread` equivalent for `child_process`. I don't think there's a way to use that without introducing new configuration option but let's leave that to another PR.

**Comment by AriPerkkio** [comment]
> [...] and implement `--single-thread` equivalent for `child_process`. I don't think there's a way to use that without introducing new configuration option but let's leave that to another PR.

Looks like I already implemented this. It's available with `vitest --no-threads --single-thread`. The `--no-thread` makes Vitest pick `pools/child.ts` and `--single-thread` makes it run in a single process. 

> #### TODO
>
> - [ ] Reporters work weird in test/public-api and test/stacktraces when run with --no-threads. Enabling CI=true fixes this.

Looks like this bug is present on `main` too. When running Vitest through JS API (`startVitest()`) with `--no-threads`, the reporters are not resolved correctly. This can be tested with:

```sh
cd vitest/test/public-api

# Pass
pnpm run test --threads --run
CI=true pnpm run test --no-threads --run

# Fail
pnpm run test --no-threads --run
```

**Comment by ghiscoding** [comment]
so is this complete? on which milestone should we expect this to land, in v0.35 or v1.0? As you can see I'm very exited about it 😄

**Comment by sheremet-va** [comment]
> so is this complete? on which milestone should we expect this to land, in v0.35 or v1.0? As you can see I'm very exited about it 😄

The next major is 1.0.0-beta.0

**Comment by ghiscoding** [comment]
> The next major is 1.0.0-beta.0

@sheremet-va ok but just so you know, you still have a v0.35 milestone that exist with 1 PR assign to it 😋 
https://github.com/vitest-dev/vitest/milestone/11

awesome work guys

**Comment by ghiscoding** [comment]
> Many projects are unable to use `node:worker_threads` and are forced to use a single `node:child_process` via `--no-threads`. That's stable, but slow. This PR allows using multiple `node:child_process` for running tests similarly as `--threads` does.
> * [Hitting Segmentation fault when running Vitest . #3143](https://github.com/vitest-dev/vitest/issues/3143)
> * [Vitest crashes when running tests with Prisma #3106](https://github.com/vitest-dev/vitest/issues/3106)
> * [Getting Node low level errors running vitest #3816](https://github.com/vitest-dev/vitest/issues/3816)

@AriPerkkio I'm just curious, were the issues you referenced supposed to be fixed (aka closed) since this PR got merged? Thanks a ton for the PR on my side 😃

**Comment by AriPerkkio** [comment]
Yes, I think we should close those now. These changes are the best solution we can offer for cases where `node:worker_threads` simply won't work with users code. 

I'll go through those later today and leave a comment describing the changes while closing them.
